docs/RFCS/20220818_tenant_version_upgrades.md

```diff
@@ -0,0 +1,475 @@
+- Feature Name: Tenant version upgrades & migrations
+- Status: draft
+- Start Date: 2022-08-18
+- Authors: knz with help from dt and ajw
+- RFC PR: [#86377](https://github.com/cockroachdb/cockroach/pull/86377)
+- Cockroach Issues: [#66606](https://github.com/cockroachdb/cockroach/issues/66606) [#80992](https://github.com/cockroachdb/cockroach/issues/80992) [CRDB-10829](https://cockroachlabs.atlassian.net/browse/CRDB-10829)
+- Jira design epic: [CRDB-14533](https://cockroachlabs.atlassian.net/browse/CRDB-14533)
+
+# Summary
+
+This RFC proposes a mechanism inside CockroachDB to orchestrate the
+initialization of system tables, then upgrading the logical cluster
+version seen by secondary tenants.
+
+This aims to fix an active correctness bug, for which we so far
+use a kludgy workaround in CC Serverless, and which will block
+a generalization of multi-tenancy to all deployments.
+
+The RFC also clarifies concepts and terminology, which we expect
+will simplify conversations around migrations and increase
+the ability of the CRL engineering team to reason and talk about
+version upgrades.
+
+<!-- markdown-toc start - Don't edit this section. Run M-x markdown-toc-refresh-toc -->
+**Table of Contents**
+
+- [Summary](#summary)
+- [Background](#background)
+    - [Introduction](#introduction)
+    - [Invariant between logical and executable versions](#invariant-between-logical-and-executable-versions)
+    - [What gets controlled by logical versions](#what-gets-controlled-by-logical-versions)
```

Comment:
the "logical version" described here widely understood as and referred to as "cluster version"; I think using a different new name for it here just adds cognitive overhead for the reader; I'd describe it _as_ a logical version when introducing it or when talking about logical behavior differences, but then just referring the version itself, it by name, I think we should just use the existing adopted terminology of "cluster version".